Bluetooth: Multiple identity support

jhedberg opened the pull request:
This set of patches adds support for multiple identities (identity addresses and local IRKs) for LE peripheral role operations. For central role the default identity (0) is always used.

Not very thoroughly tested yet (hence the RFC/DNM tag).

Fixes #7473

Commit:
Bluetooth: Minor whitespace fixes

Remove redundant whitespace.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Add infrastructure to handle multiple identities

Make it possible to have multiple identity addresses as an LE
peripheral. For central role only the default identity is supported
for now. This also extends the flash storage in a backward compatible
way.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Add initial APIs for identity manipulation

Add APIs for getting current identities as well as for creating new
ones.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Remove redundant BT_DEV_ID_STATIC_RANDOM flag

This flag was both redundant and creates confusion with potentially
multiple identity addresses (it was only referring to the first
identity address). We can simply just look at the type of the identity
address wanting to be used.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Add full storage support for multiple identities

Update the storage handling to take into account multiple identities.
We can save a bit of code by using the new bt_id_create() API from
within settings.c.

Also make the treatment of addr & irk parameters to bt_id_create()
consistent, in that NULL is acceptable for both of them.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: shell: Add multiple identity support

Add shell support for creating and showing identities.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

pizi-nordic commented:
FYI: @pdunaj

pdunaj commented:
FYI @fiku

pdunaj commented:
@jhn-nordic

pizi-nordic commented on include/bluetooth/bluetooth.h:
```diff
@@ -71,6 +79,36 @@ int bt_enable(bt_ready_cb_t cb);
+ *  @param irk  Identity Resolving Key (16 bytes) to be used with this
+ *              identity. If set to all zeroes or NULL, the stack will
+ *              generate a random IRK for the identity and copy it back
+ *              to the parameter upon return from this function. If
```
IMHO we should not copy-back IRK if *irk is NULL :)

jhedberg replied:
That's what the implementation does :) I'll fix the documentation.

pizi-nordic commented on include/bluetooth/bluetooth.h:
```diff
@@ -29,6 +29,14 @@
+ *  make the code more readable, especially when only one identity is
+ *  supported.
+ */
+#define BT_ID_DEFAULT 0
```
Is default identity always available or have to be generated using bt_id_create()? If it is always available, please state so in the comment.

jhedberg replied:
It can be created, if bt_id_create() is called before bt_enable(). Also, the settings storage will call bt_id_create() to create the default identity if one wasn't found in flash and the controller has no public address. I'll try to improve the documentation regarding this.

pizi-nordic commented on include/bluetooth/bluetooth.h:
```diff
@@ -71,6 +79,36 @@ int bt_enable(bt_ready_cb_t cb);
+ *  @return Identity identifier (>= 0) in case of success, or a negative
+ *          error code on failure.
+ */
+int bt_id_create(bt_addr_le_t *addr, u8_t *irk);
```
How we can delete identity?

jhedberg replied:
In the github issue that tracks this (#7473) there seemed to be a consensus that this would not be a safe API to have. So the only way to delete stored identities is to erase the flash storage area and do a power cycle.

pizi-nordic replied:
I don't think that it is acceptable solution. Please look for use cases for multiple identities:
https://github.com/zephyrproject-rtos/zephyr/issues/7473#issuecomment-388741527

We definitely must have a method for deleting (or revoking and later recycling) given identity.

jhedberg replied:
With `bt_id_reset()` that I added you can both revoke and recycle. To revoke you'd simply issue `bt_unpair(id, NULL);` and to recycle you'd call `bt_id_reset()`, or in a single step just call `bt_id_reset()`.

pizi-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -350,6 +350,14 @@ config BT_DEVICE_APPEARANCE
 
+config BT_ID_MAX
+	int "Maximum number of local identities"
+	range 1 36
```
Why 36? (I am just curious.)

jhedberg replied:
The current settings implementation has a limit of 256 byte values. bt_addr_le_t is 7 bytes and 36 * 7 = 252. That said, I added this before considering IRKs, which are 16 bytes. Also, I failed to consider the fact that there's base64 encoding before writing to flash. So this current maximum is not what you can reach in practice. Most likely it's closer to 10 or so. That said, there's ongoing work to remove the 256 byte value limit from settings. It'd be nice if Kconfig would allow giving only a minimum or maximum but leave the other limit unbounded.

pizi-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -893,9 +893,10 @@ static int start_security(struct bt_conn *conn)
 		if (!conn->le.keys) {
 			conn->le.keys = bt_keys_find(BT_KEYS_LTK_P256,
-						     &conn->le.dst);
+						     conn->id, &conn->le.dst);
```
Could we be consistent and use the same pattern for all calls? (ID as first argument).

jhedberg replied:
The pattern I've tried to follow is that the identity precedes remote addresses, so in that sense this is consistent. The general idea is that a remote address no longer uniquely identifies a remote address, rather the tuple (local identity, remote address) must be used. Because of this I'd keep these two parameters together in the parameter lists.

pizi-nordic replied:
Ok.

carlescufi commented:
CC @pdunaj @Qbicz @jhn-nordic

carlescufi commented:
@aescolar does anyone from Oticon want to review this?

pizi-nordic reviewed (commented):
General question: Are we going to support per-identity values in GATT database different than CCCD?

jhedberg commented on include/bluetooth/bluetooth.h:
```diff
@@ -71,6 +79,36 @@ int bt_enable(bt_ready_cb_t cb);
+ *  Returns an array of the currently configured identity addresses.
+ *
+ *  @param addrs Array where to store the configured identities.
+ *  @param count Should be initialized to the array size. Once the function
```
Use an array with CONFIG_BT_ID_MAX entries.

jhedberg replied:
I realize that it's less useful to want to get anything else than this amount, but at least the API would allow getting the default identity (using one element array), if the app doesn't care for the others.

pizi-nordic replied:
Are you planning to have API allowing get the most/least used identity or iterate over identities in some predefined order?

Usecases:
- Direct advertising to multiple masters (starting from the most frequently used identity). 
- Remove the least used identity in order to make room for new.

jhedberg replied:
> Are you planning to have API allowing get the most/least used identity or iterate over identities in some predefined order?

I would leave this up to the application. General principle being: keep the API as simple as possible and only add new features to it if they're not possible to do with the existing API plus some extra logic on the application side.

carlescufi commented:
@pizi-nordic 
>General question: Are we going to support per-identity values in GATT database different than CCCD?

Well, we get read/write callbacks with `struct bt_conn *conn`. At that point the application can check with `id` is active and reply with a data set or another?
@jhedberg can you confirm my statement?

jhedberg commented:
> Well, we get read/write callbacks with `struct bt_conn *conn`. At that point the application can check with `id` is active and reply with a data set or another? @jhedberg can you confirm my statement?

@carlescufi yes, that would be the way to do it.

jhedberg commented:
@pizi-nordic I've tried to cover the issues you raised in the API documentation. I've also reduced the maximum to 10 identities (in theory current settings subsystem would allow up to 11, but I decided to keep it a nice round number)

oyvindronningstad commented on include/bluetooth/bluetooth.h:
```diff
@@ -71,6 +79,56 @@ int bt_enable(bt_ready_cb_t cb);
+
+/** @brief Create a new identity.
+ *
+ *  Create a new identity using the given address and IRK. This function
```
Just looking at the API docs, I'd suggest clarifying the difference between calling this before and after bt_enable(). Maybe they should even be two separate functions?

jhedberg replied:
Separate functions seems a bit excessive (I have also no idea how to name them intuitively in that case). I do have at least some documentation on the differences in usage, I wonder if you were looking at an earlier revision of the patch set?

jhedberg replied:
> "can be used to override the controller's public address" does this mean I have to do more to make this happen, or will it always happen.

It will always happen. It would get overwritten though if there was something stored in flash and the app calls `settings_load()`. Btw, the design is like this to keep the existing `bt_set_id_addr()` functionality which was added for Zephyr 1.12. Now that API is just a wrapper for `bt_id_create()`.

> "(in case it has one)" what happens if it doesn't?

If the controller has no public address, then `bt_id_create()` can be used before bt_enable() to populate pre-defined identity addresses, starting with the default identity. The stack will auto-generate a static random address for the default identity if one is not found in flash and was also not provided through bt_id_create().

> Is the id returned from a call before bt_enable() equivalent to ids returned after bt_enable() until next reset? I.e. can it be used for all the same things?

Yes, the id is the same, it's just an array index to the stack-internal array of identities. Only difference is that before bt_enable() nothing will get written to flash, i.e. if an app does this, then the app needs to repeat the same `bt_id_create()` call upon every bootup.

> Are _ANY/NULL parameters valid before bt_enable()?

Yes. No difference there.

> Does the controller's default public address have an ID?

Yes, it will be BT_ID_DEFAULT (i.e. 0) if available and not overridden by calling `bt_id_create()` before `bt_enable()`.

jhn-nordic replied:
Can the identity address be changed with bt_set_id_addr()? With the limit of 10 identies we will run out of identites quite fast if we do not recycle old ones. I.e in the mouse use case where you have a mouse that can support 3 identies. What happens when we want to add a new bond? We can either create a new identity with a new address (will reach 10 fast) or can we change the address for an existing identity.

jhedberg replied:
I'd then consider adding something like the following:
```
int bt_id_reset(u8_t id, bt_addr_t *addr, u8_t *irk);
```

This would reinitialize the identity and remove any bondings it had.

The alternative approach would be to modify `bt_id_create()`:
```
#define BT_ID_ANY 0xff

int bt_id_create(u8_t id, bt_addr_t *addr, u8_t *irk);
```

If `id` is anything else than BT_ID_ANY it would need to be an already created id, otherwise the function would return failure. When given an already created id this would behave the same way as my `bt_id_reset()` proposal.

oyvindronningstad replied:
So `bt_id_create()` will override ids in some instances but not others? What happens if persistent identities are created in between non-persistent ids? Do they get consecutive ids?

jhedberg replied:
It will override if you give it a valid id. I guess BT_ID_NEW would be more intuitive than BT_ID_ANY in my example.

Mixing persistent (stack-generated address) and non-persistent (app-generated address) IDs is something I didn't think about. The API as drafted currently would allow it, but I fail to imagine a good use case for it. Do you have one in mind? The way the addresses and IRKs are stored in flash is using the stack-internal array as-is. I.e. with the current implementation any persistent id would cause all ids to be written to flash.

oyvindronningstad replied:
>Do you have one in mind?

Not really, but users will use it in all sorts of ways, so it might be good to document it.

Well maybe there is one:

 1. Call bt_id_create() to override controller's address
 2. bt_enable()
 3. settings_load()
 4. Call bt_id_create() to generate random ids.

Will 4. cause 1. to be written to flash?

oyvindronningstad replied:
How about adding a

    bt_id_store()

Which stores to flash whatever would be returned by `bt_id_get()` at that time.

jhedberg replied:
There are a few issues I see with that:

 - We don't have explicit "store" APIs for anything else in the stack. Information (pairing keys, GATT CCC, etc) is stored implicitly
 - You risk getting into a situation where the app doesn't call `bt_id_store()` but the stack goes and performs a pairing and ends up storing information related to the id, but not the id itself.

jhedberg replied:
I'd like to add that currently the storing of ids is deferred through the system work queue, i.e. the app can do multiple bt_id_create(); bt_id_create(); ... calls but it should still just result in just a single write to flash.

oyvindronningstad replied:
> We don't have explicit "store" APIs for anything else in the stack. Information (pairing keys, GATT CCC, etc) is stored implicitly

I see

> You risk getting into a situation where the app doesn't call bt_id_store() but the stack goes and performs a pairing and ends up storing information related to the id, but not the id itself.

Can't this situation already happen? Not all ids are stored persistently, and the stack can't really know which is which.

jhedberg replied:
> Can't this situation already happen? Not all ids are stored persistently, and the stack can't really know which is which.

What the stack assumes is that if it is given pre-defined values, then the app is responsible for providing these always. They might be hard-coded into the app (this is the use case why `bt_set_id_addr()` was added), come from some factory-programmed register, special flash location or something else. So it would be an application bug if it fails to keep providing these ids for every power cycle. I can try to make this clear in the documentation.

oyvindronningstad replied:
> What the stack assumes is that if it is given pre-defined values, then the app is responsible for providing these always.

I see. This gets a bit more complicated when considering DFU updates, though. IMO the user should have the possibility to store even predefined identities if that is more convenient for them.

jhedberg replied:
Would it be an option then to always store the identities post-bt_enable, regardless of being stack-generated or pre-defined? Before bt_enable I wouldn't touch flash however, since at that point we don't know the flash content and would possibly be (over)writing blindly to flash.

oyvindronningstad replied:
As a user, I think I would find that less confusing. Maybe a Kconfig could contol it? I don't know the conventions.

oyvindronningstad commented on include/bluetooth/bluetooth.h:
```diff
@@ -71,6 +79,56 @@ int bt_enable(bt_ready_cb_t cb);
+ *  If an insufficient amount of identities were recovered the app may then
+ *  call bt_id_create() to create new ones.
+ *
+ *  @param addr Address to use for the new identity. If initialized to
```
I think it's less confusing if generating a random value looks the same for addr and irk. So NULL addr should generate a random value, and there could be a BT_IRK_LE_ANY defined to all zeroes.

jhedberg replied:
Yep, makes sense. I'll fix this in the next revision.

jhedberg replied:
Actually, doing this for the IRK seems a bit excessive. The documentation asks to use all zeroes, which is fairly straight-forward to do in the app, i.e. `u8_t irk[16] = { 0 };`. If I went with a define similar to BT_ADDR_LE_ANY one would need to do a memcpy separate from the variable declaration/initialization, which is less efficient. I've fixed this at least for the address in the latest revision.

oyvindronningstad replied:
> If I went with a define similar to BT_ADDR_LE_ANY one would need to do a memcpy separate from the variable declaration/initialization, which is less efficient.

I'm not sure this is correct, BT_ADDR_LE_ANY is a pointer, which BT_IRK_LE_ANY could also be without having to do any copying.

    #define BT_ADDR_LE_ANY  (&(bt_addr_le_t) { 0, { { 0, 0, 0, 0, 0, 0 } } }) // current definition
    #define BT_IRK_LE_ANY   ((u8_t[16]) { 0 })
    ...
    u8_t * irk1   = BT_IRK_LE_ANY; // works
    u8_t irk2[]   = BT_IRK_LE_ANY; // doesn't work (non-constant initializer)
    u8_t irk3[16] = BT_IRK_LE_ANY; // doesn't work (non-constant initializer)
    foo(BT_IRK_LE_ANY);            // works (foo takes u8_t pointer argument)

jhedberg replied:
I understand what you mean. I'm not sure we should encourage passing the *_ANY macros directly as input though, since the result is the same as passing NULL, except that the caller will have no way to retrieve the value afterward while still having consumed the memory on the stack to store it. `u8_t *irk = BT_IRK_ANY;` would indeed be a sane thing, with a slightly heavier memory footprint due to pointer (4 bytes) + 16 bytes instead of just 16 bytes. If there's a general consensus that the macro does improve the resulting code, then I'm not opposed to adding it, but it still feels a bit "low value" to me.

oyvindronningstad replied:
Sure, we can wait and see if it makes sense.

Commit:
Bluetooth: Pass identity to connection and pairing clearing functions

When doing bt_unpair() we need to pass the given identity when
disconnecting and clearing keys, in case all associated pairings were
requested to be cleared.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Introduce bt_id_reset() API

Add a new API which can be used to reclaim an identity slot for a new
identity. When called, any previous pairings, connections, or other
data will be cleared, and then a new identity will be generated in the
place of the old one.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
Updated the PR with the new bt_id_reset() API (untested and no shell support yet). Also simplified the persistent storage logic as discussed, i.e. after bt_enable storing will always happen. We can introduce a Kconfig option later to modify this behavior, if it's deemed to be necessary.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2217,8 +2220,16 @@ int bt_gatt_store_ccc(const bt_addr_le_t *addr)
 
-	bt_settings_encode_key(key, sizeof(key), "ccc",
-			       (bt_addr_le_t *)addr, NULL);
+	if (id) {
```
The idea is to keep the storage backwards and forwards compatible, i.e. so you're able to upgrade a Zephyr 1.12 firmware to 1.13, or vice-versa (in which case the default identity will work and others will be ignored). We're also saving a little bit of flash for the common case (single id supported) with the shorter settings path.

jhedberg commented on include/bluetooth/bluetooth.h:
```diff
@@ -71,6 +79,89 @@ int bt_enable(bt_ready_cb_t cb);
+ *  @return Identity identifier (>= 0) in case of success, or a negative
+ *          error code on failure.
+ */
+int bt_id_reset(u8_t id, bt_addr_le_t *addr, u8_t *irk);
```
Wouldn't the application anyway need to track unused identities, i.e. know what it's called `bt_id_clear()' for, so it doesn't go calling bt_le_adv_start() and triggering errors, or which identities it needs to call `bt_id_reset()` for before being able to use it again. So I'm not completely buying this argument. Another reason I'm reluctant to adding this is that it will make the implementation more complex and the storage less efficient in that there can be "holes" in the arrays that are stored. API-wise, this is not something that has an impact on what's in the current PR, so I hope this will not prevent proceeding with the rest.

Vudentz commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4851,6 +4851,41 @@ void bt_id_get(bt_addr_le_t *addrs, size_t *count)
+	}
+}
+
 int bt_id_create(bt_addr_le_t *addr, u8_t *irk)
```
It doesn't look like you are looking up for duplicated addresses, in case the application already assigned the same address to another instance, I guess that should be considered an invalid use of the API and return an error if that happens.

jhedberg replied:
I thought about this, but I'm not sure it's worth increasing the code size for what would essentially be an application bug. The identities are identified by the array index, so providing the same address would not confuse the stack in any way (it may confuse other devices though since e.g. pairing keys would only be available on one of the same-address identities)

jhedberg replied:
@Vudentz after thinking even more about this, I went and introduced duplicate identity address guards. We need to check for these either way when generating new random addresses (the chance for a collision is small but still exists).

Commit:
Bluetooth: Add safe-guards for creating duplicate identities

Make sure the application doesn't pass existing identity addresses to
bt_id_create() and bt_id_reset(). Also make sure we don't accidentally
create a duplicate when generating random identity addresses.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Introduce bt_id_delete() API

This API makes it possible to delete an existing identity and to flag
its storage slot as unused.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: shell: Add id-delete command

Add a command to test the new bt_id_delete() API.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@pizi-nordic there's now a bt_id_delete() API. After thinking more about this I agree that makes sense, since that way the app doesn't need to do any of its own persistent storage for knowing what identities are valid.

jhedberg commented:
@pizi-nordic @oyvindronningstad any further comments on this PR? I've tried to address all issues brought up so far.

pizi-nordic commented:
@jhedberg: Looks good for me.

Commit:
Bluetooth: Remove bogus bt_keys_func_t typedef

This wasn't used anywhere and was typed incorrectly (the foreach
callback takes two parameters). There was also one user of this which
was triggering compiler warnings of mismatched callback type.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Remove return parameter from bt_id_add/del

None of the callers of these APIs do anything with the return value,
so just remove it to produce more efficient code.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
I removed the DNM label now, since this seems to be working quite well in my own testing. Next it'd be nice to get some "approved" reviews :)

dbkinder reviewed (commented):
no docs to review?

jhedberg commented:
> no docs to review?

@dbkinder well, there's the Kconfig help text and the API documentation in the header files.

Vudentz reviewed (approved):
LGTM
